// dotenv-extended/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.load = exports.config = exports.parse = undefined;

var _dotenv = require('dotenv');

var _dotenv2 = _interopRequireDefault(_dotenv);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Created by Keith Morris on 2/9/16.
 */
function loadEnvironmentFile(path, encoding, silent) {
    try {
        var data = _fs2.default.readFileSync(path, encoding);
        return _dotenv2.default.parse(data);
    } catch (err) {
        if (!silent) {
            console.error(err.message);
        }
        return {};
    }
}
var parse = exports.parse = _dotenv2.default.parse.bind(_dotenv2.default);
var config = exports.config = function config(options) {

    var defaultsData,
        environmentData,
        defaultOptions = {
        encoding: 'utf8',
        silent: true,
        path: '.env',
        defaults: '.env.defaults',
        schema: '.env.schema',
        errorOnMissing: false,
        errorOnExtra: false,
        errorOnRegex: false,
        includeProcessEnv: false,
        assignToProcessEnv: true,
        overrideProcessEnv: false
    };

    options = Object.assign({}, defaultOptions, options);

    defaultsData = loadEnvironmentFile(options.defaults, options.encoding, options.silent);
    environmentData = loadEnvironmentFile(options.path, options.encoding, options.silent);

    var configData = Object.assign({}, defaultsData, environmentData);

    if (options.errorOnMissing || options.errorOnExtra || options.errorOnRegex) {
        var schema = loadEnvironmentFile(options.schema, options.encoding, options.silent);
        var _config = options.includeProcessEnv ? Object.assign({}, configData, process.env) : configData;
        var schemaKeys = Object.keys(schema);
        var configKeys = Object.keys(_config);

        var missingKeys = schemaKeys.filter(function (key) {
            return configKeys.indexOf(key) < 0;
        });
        var extraKeys = configKeys.filter(function (key) {
            return schemaKeys.indexOf(key) < 0;
        });
        if (options.errorOnMissing && missingKeys.length) {
            throw new Error('MISSING CONFIG VALUES: ' + missingKeys.join(', '));
        }

        if (options.errorOnExtra && extraKeys.length) {
            throw new Error('EXTRA CONFIG VALUES: ' + extraKeys.join(', '));
        }

        if (options.errorOnRegex) {
            var regexMismatchKeys = schemaKeys.filter(function (key) {
                if (schema[key]) {
                    return !new RegExp(schema[key]).test(_config[key]);
                }
            });

            if (regexMismatchKeys.length) {
                throw new Error('REGEX MISMATCH: ' + regexMismatchKeys.join(', '));
            }
        }
    }

    if (options.assignToProcessEnv) {
        if (options.overrideProcessEnv) {
            Object.assign(process.env, configData);
        } else {
            var tmp = Object.assign({}, configData, process.env);
            Object.assign(process.env, tmp);
        }
    }
    return configData;
};

var load = exports.load = config;

exports.default = { parse: parse, config: config, load: load };
